packages/block-editor/src/store/selectors.js

```diff
@@ -1259,15 +1261,36 @@ const canInsertBlockTypeUnmemoized = (
 		parentName
 	);
 
-	if ( hasParentAllowedBlock !== null && hasBlockAllowedParent !== null ) {
-		return hasParentAllowedBlock || hasBlockAllowedParent;
-	} else if ( hasParentAllowedBlock !== null ) {
-		return hasParentAllowedBlock;
-	} else if ( hasBlockAllowedParent !== null ) {
-		return hasBlockAllowedParent;
+	if (
+		hasParentAllowedBlock !== null &&
+		hasBlockAllowedParent !== null &&
+		! hasParentAllowedBlock &&
+		! hasBlockAllowedParent
+	) {
+		return false;
+	} else if ( hasParentAllowedBlock !== null && ! hasParentAllowedBlock ) {
+		return false;
+	} else if ( hasBlockAllowedParent !== null && ! hasBlockAllowedParent ) {
+		return false;
```

Comment:
Why we are changing the logic here if we are planning on using a filter? We want to apply the filter only in case the allowed blocks at this stage?

Reply:
That was my thinking, the filter allows you to change `true` to `false`. Otherwise, it would have to run on every `return` path, and there's plenty of them.

Reply:
I just simplified this logic further, should be easier to read now.